airbyte-integrations/connectors/source-bing-ads/source_bing_ads/source.py

```diff
@@ -0,0 +1,302 @@
+    ) -> MutableMapping[str, Any]:
+        return {"CampaignId": stream_slice["campaign_id"], "ReturnAdditionalFields": self.additional_fields}
+
+    def stream_slices(
+        self,
+        sync_mode: SyncMode,
+        cursor_field: List[str] = None,
+        stream_state: Mapping[str, Any] = None,
+    ) -> Iterable[Optional[Mapping[str, Any]]]:
+        campaigns = Campaigns(self.client, self.config)
+        for account_id in self.config["account_ids"]:
+            for campaign in campaigns.read_records(sync_mode=SyncMode.full_refresh, stream_slice={"account_id": account_id}):
+                yield {"campaign_id": campaign["Id"], "account_id": account_id}
+
+        yield from []
+
+
+class Ads(BingAdsStream):
+    """
+    Retrieves the ads for all provided accounts.
+    API doc: https://docs.microsoft.com/en-us/advertising/campaign-management-service/getadsbyadgroupid?view=bingads-13
+    Ad schema: https://docs.microsoft.com/en-us/advertising/campaign-management-service/ad?view=bingads-13
+    """
+
+    data_field: str = "Ad"
+    service_name: str = "CampaignManagement"
+    operation_name: str = "GetAdsByAdGroupId"
+    additional_fields: str = "ImpressionTrackingUrls Videos LongHeadlines"
+    ad_types = ["Text", "Image", "Product", "AppInstall", "ExpandedText", "DynamicSearch", "ResponsiveAd", "ResponsiveSearch"]
+
+    def request_params(
+        self,
+        stream_state: Mapping[str, Any],
+        stream_slice: Mapping[str, Any] = None,
+        next_page_token: Mapping[str, Any] = None,
+    ) -> MutableMapping[str, Any]:
+        return {
+            "AdGroupId": stream_slice["ad_group_id"],
+            "AdTypes": {"AdType": self.ad_types},
+            "ReturnAdditionalFields": self.additional_fields,
+        }
+
+    def stream_slices(
+        self,
+        sync_mode: SyncMode,
+        cursor_field: List[str] = None,
+        stream_state: Mapping[str, Any] = None,
+    ) -> Iterable[Optional[Mapping[str, Any]]]:
+        ad_groups = AdGroups(self.client, self.config)
+        for slice in ad_groups.stream_slices(sync_mode=SyncMode.full_refresh):
+            for ad_group in ad_groups.read_records(sync_mode=SyncMode.full_refresh, stream_slice=slice):
+                yield {"ad_group_id": ad_group["Id"], "account_id": slice["account_id"]}
+        yield from []
+
+
+class SourceBingAds(AbstractSource):
+    def check_connection(self, logger: AirbyteLogger, config: Mapping[str, Any]) -> Tuple[bool, Any]:
+        try:
+            client = Client(**config)
+            client.get_service(service_name="CustomerManagementService").GetAccount(AccountId=config["account_ids"][0])
+        except Exception as error:
+            return False, error
+
+        return True, None
+
+    def streams(self, config: Mapping[str, Any]) -> List[Stream]:
+        client = Client(**config)
+        return [Accounts(client, config), Campaigns(client, config), AdGroups(client, config), Ads(client, config)]
```

Comment:
In one of the last Airbyte reviews, I got comments for sort this alphabetically. I think to do that before a comment from their side)